### PR TITLE
Be#37 error code

### PR DESCRIPTION
**Pull request by scotte216**
Created JSON error codes in responses and added lots of tests, including mocks for special-case exceptions.

### CODE REVIEW

File: src/cassy.py [hunk 1]
```diff
@@ -50,15 +54,18 @@ def get_env_data(node_id, env_variable):
         )
 
         if not rows:
-            raise Exception('Invalid Environmental Variable or Node ID')
+            raise Exception('No environmental data found for ' + node_id + ' and ' + env_variable + '.')
```
Comment (logicplace) [inline src/cassy.py]: Not a good idea to raise Exception (I can explain this if needed but it seems like you know and just missed a couple)

File: src/cassy.py [hunk 2]
```diff
@@ -50,15 +54,18 @@ def get_env_data(node_id, env_variable):
...
             if env_variable == 'temperature':
                 return rows[0].temperature
             elif env_variable == 'humidity':
                 return rows[0].humidity
-            else:
+            elif env_variable == 'leafwetness':
                 return rows[0].leafwetness
+            else:
+                # Shouldn't get to this point, but here for completeness
+                raise Exception('Unknown environmental variable: ' + env_variable)
```
Comment (logicplace) [inline src/cassy.py]: Here too

File: src/login/views.py [hunk 3]
```diff
@@ -30,28 +29,37 @@ def index(request):
...
+    token_object = {'token': token}
 
     try:
         # Get stored password from database, and verify with password arg
         logger.info('Fetching password for user \'' + username + '\'.')
         stored_password = cassy.get_user_password(username)
         logger.info('Successfully fetched password for user \''
                     + username + '\'.'
-        )
+                    )
```
Comment (logicplace) [inline src/login/views.py]: ???

Comment (scotte216) [inline src/login/views.py]: Donno how that snuck in there.

File: src/tests/tests.py [hunk 4]
```diff
@@ -48,6 +48,16 @@ def test_response_invalid_username(self):
         )
         self.assertEqual(response.status_code, 403)
 
+    def test_response_invalid_username_missing(self):
+        setup_test_environment()
+        client = Client()
+        response = client.get(
```
Comment (logicplace) [inline src/tests/tests.py:56]: Can you make this be post? As I mentioned in the doc, URLs aren't encrypted over HTTPS, so putting bare creds there is extremely insecure.

Comment (scotte216) [inline src/tests/tests.py:56]: We can (and should!) though probably better to make that change in a separate issue/PR.

**Comment by logicplace** [comment]
A couple of the things I just commented on once but there are a few instance.

Currently the codes seem categorical rather than 1-to-1 and I'm not sure how useful that is. It really doesn't seem useful at all to me off hand but idk. I'd be good with using your system and just adding on things to the key like you do with the "header" values in that dict. Whether that's numbers or more specific text. I don't know if I care as much for unknowns, though, might just be good to log a stack trace in those cases.

Love how many tests there are wow

File: src/tests/tests.py (hunk 4 above)
Comment (logicplace) [inline src/tests/tests.py:56]: It seems wrong to test for the wrong functionality, even in that case.

Comment (SoySauceLabs) [inline src/tests/tests.py:56]: Yes, I meant to change this to a post on my branch, although it seems I created my PR without it...I'll add it in. :)

**Comment by scotte216** [comment]
I can make it more 1-to-1. I was trying to strike a balance for the type of errors while leaving the flexibility to add additional information if desired. I was trying to avoid making a different exception for every scenario (or for most scenarios), but maybe that's silly of me! These kinds of changes should be pretty straight forward to make so I can do it. 

The only one I want to keep somewhat vague are those related to username/password so one couldn't glean information based on the error message. 

Who do we envision as the primary user of these messages? Visitors or developers? For example, a visitor would probably have no idea what to do with something like "unable to get auth token from database" whereas a developer would find that very helpful. :)

**Comment by logicplace** [comment]
That's fine to me, I more meant code <-> message being 1-to-1ish, not code <-> error location. In terms of exceptions, having only a few exceptions is fine for the python-side, tho you can't distinguish "Exception"s we generate from internally generated issues.

Errors returned are more meant for end users. Seems like there's separate logging for devs, anyway. Though I think, even if they don't know what a database in, it's good to sort of situate the error as something that they'd have to contact the devs about. Tho instructions like "this has been logged" or "contact support if this keeps happening" or "try again" may be worthwhile.

**Comment by scotte216** [comment]
KERPOW! Okay, refactored error codes. The returned text is easily modifiable in a centralized location (src/common/errors.py) and they are much more specific now (both the codes and the messages). ALSO the tests now check the returned error codes.

**Comment by logicplace** [comment]
If I have to accept the tests with all the auth stuff in the URL then it looks good to me. But it feels like pandering to coveralls.

I'll test this soon before merging/closing.

**Comment by scotte216** [comment]
I made a slight modification to have all the error codes be centralized so it'll be easier to change them should anyone want to rename them.

**Comment by logicplace** [comment]
One thing to note is that if arguments aren't specified it currently does not throw a 400 error. I'd guess it's assuming them as empty strings. That's fine for now but it should be changed in the future. I should probably open a ticket for it. But otherwise lgtm!

**Pull request merged**